File: MyApplication/library/src/main/java/com/a65apps/kalimullinilnazrafilovich/library/applicaiton/presenters/RouteMapPresenter.java
package com.a65apps.kalimullinilnazrafilovich.library.applicaiton.presenters;

import androidx.annotation.NonNull;

import com.a65apps.kalimullinilnazrafilovich.entities.Point;
import com.a65apps.kalimullinilnazrafilovich.interactors.location.ContactLocationInteractor;
import com.a65apps.kalimullinilnazrafilovich.interactors.route.RouteInteractor;
import com.a65apps.kalimullinilnazrafilovich.library.applicaiton.views.RouteMapView;
import com.google.android.gms.maps.model.LatLng;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.schedulers.Schedulers;
import moxy.InjectViewState;
import moxy.MvpPresenter;

@InjectViewState
public class RouteMapPresenter extends MvpPresenter<RouteMapView> {
    private final transient ContactLocationInteractor contactLocationInteractor;
    private final transient RouteInteractor routeInteractor;
    private final transient CompositeDisposable compositeDisposable;

    public RouteMapPresenter(@NonNull ContactLocationInteractor contactLocationInteractor,
                             @NonNull RouteInteractor routeInteractor) {
        compositeDisposable = new CompositeDisposable();

        this.contactLocationInteractor = contactLocationInteractor;
        this.routeInteractor = routeInteractor;
    }

    public void showMarkers() {
        compositeDisposable
                .add(contactLocationInteractor.loadAllContactLocations()
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(
                                (list) -> getViewState().showMarkers(list),
                                Throwable::printStackTrace
                        ));
    }

    public void showRoute(@NonNull Point from, @NonNull Point to) {
        compositeDisposable
                .add(routeInteractor.loadRoute(from, to)
                        .subscribeOn(Schedulers.io())
                        .map((route) -> getListLatLng(routeInteractor.routeToPoints(route)))
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(
                                (dots) -> {
                                    if (dots.isEmpty()) {
                                        getViewState().showMessageNoRoute();
                                    } else {
                                        getViewState().showRoute(dots);
                                    }
                                },
                                Throwable::printStackTrace
                        )
                );
    }

    @NonNull
    private List<LatLng> getListLatLng(@NonNull List<Point> points) {
        if (points.isEmpty()) {
            return new ArrayList<>();
        } else {
            List<LatLng> latLngs = new ArrayList<>();
            for (Point point : points) {
                latLngs.add(new LatLng(point.getLatitude(), point.getLongitude()));
            }
            return latLngs;
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        compositeDisposable.dispose();
    }
}

File: library/src/main/java/com/a65apps/kalimullinilnazrafilovich/library/applicaiton/presenters/RouteMapPresenter.kt
package com.a65apps.kalimullinilnazrafilovich.library.applicaiton.presenters

import android.util.Log
import com.a65apps.kalimullinilnazrafilovich.entities.Point
import com.a65apps.kalimullinilnazrafilovich.entities.Route
import com.a65apps.kalimullinilnazrafilovich.interactors.location.ContactLocationInteractor
import com.a65apps.kalimullinilnazrafilovich.interactors.route.RouteInteractor
import com.a65apps.kalimullinilnazrafilovich.library.applicaiton.views.RouteMapView
import com.google.android.gms.maps.model.LatLng
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.schedulers.Schedulers
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.launch
import moxy.InjectViewState
import moxy.MvpPresenter
import java.util.*
import javax.inject.Inject

@InjectViewState
class RouteMapPresenter @Inject constructor(
        private val contactLocationInteractor: ContactLocationInteractor,
        private val routeInteractor: RouteInteractor
) : MvpPresenter<RouteMapView>() {
    private val compositeDisposable = CompositeDisposable()

    private lateinit var jobMarkers: Job

    fun showMarkers() {
        try {
            jobMarkers = CoroutineScope(Dispatchers.Main).launch {
                contactLocationInteractor.loadAllContactLocations()
                        .flowOn(Dispatchers.IO)
                        .collect { list ->
                            viewState.showMarkers(list)
                        }
            }
        } catch (e: Exception) {
            Log.e(this::class.simpleName, e.printStackTrace().toString())
        }
    }

    fun showRoute(from: Point, to: Point) {
        compositeDisposable
                .add(routeInteractor.loadRoute(from, to)
                        .subscribeOn(Schedulers.io())
                        .map { route: Route? -> getListLatLng(routeInteractor.routeToPoints(route)) }
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(
                                { dots: List<LatLng> ->
                                    if (dots.isEmpty()) {
                                        viewState.showMessageNoRoute()
                                    } else {
                                        viewState.showRoute(dots)
                                    }
                                }) { obj: Throwable -> obj.printStackTrace() }
                )
    }


    private fun getListLatLng(points: List<Point>): List<LatLng> {
        return if (points.isEmpty()) {
            ArrayList()
        } else {
            val latLngs = mutableListOf<LatLng>()
            for ((latitude, longitude) in points) {
                latLngs.add(LatLng(latitude, longitude))
            }
            latLngs
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        compositeDisposable.dispose()
        jobMarkers.cancel()
    }
}